pkg/wireguard/agent/agent.go

```diff
@@ -0,0 +1,356 @@
+// Copyright 2021 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This package contains the agent code used to configure the Wireguard tunnel
+// between nodes. The code supports adding and removing peers at run-time
+// and the peer information is retrieved via the CiliumNode object.
+package agent
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"strconv"
+
+	"github.com/cilium/cilium/pkg/cidr"
+	"github.com/cilium/cilium/pkg/datapath/linux/linux_defaults"
+	"github.com/cilium/cilium/pkg/datapath/linux/route"
+	"github.com/cilium/cilium/pkg/lock"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/cilium/cilium/pkg/node"
+	"github.com/cilium/cilium/pkg/option"
+	"github.com/cilium/cilium/pkg/sysctl"
+	"github.com/cilium/cilium/pkg/wireguard/types"
+
+	"github.com/sirupsen/logrus"
+	"github.com/vishvananda/netlink"
+	"golang.org/x/sys/unix"
+	"golang.zx2c4.com/wireguard/wgctrl"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+const (
+	listenPort = 51871
+)
+
+var log = logging.DefaultLogger.WithField(logfields.LogSubsys, "wireguard")
+
+type Agent struct {
```

Comment:
How about adding a few lines as a comment about how is this supposed to work?

E.g., 

> Agent is initialized with .Init() where the wireguard link will be created and the proper routes set. During Init, existing peer keys are placed into `restoredPubKeys`. Once `RestoreFinished` is called obsolete keys and peers are removed. `UpdatePeer` updates the public key of a peer.

(The above might not be fully correct -- it was my understanding from reading the code)

Reply:
Good and correct summary! Thanks, I've adapted a version of it.